onnxruntime/core/providers/op_kernel_type_control.h

```diff
@@ -9,22 +9,38 @@
 #include "boost/mp11.hpp"
 
 #include "core/common/type_list.h"
+#include "core/common/type_set_utils.h"
+
 #include "core/framework/data_types.h"
 
 /**
  * These utilities provide a way to control what types are enabled for an Op kernel implementation.
+ * At a high level, we have the notion of supported, required, allowed, and enabled type sets.
+ *
+ * 1. Supported types are the types that the Op kernel implementation supports by default.
+ *    This type set must be provided.
+ *
+ * 2. Required types are the types that are always supported.
+ *    This type set is optional. If not given, the required type set is empty.
+ *    They should be a subset of the supported types (1).
  *
- * At a high level, we have the notion of supported, allowed, and enabled types.
- * - Supported types are the types that the Op kernel implementation supports by default.
- * - Allowed types are the types for which support is requested (for example, by external configuration).
- * - Enabled types are the types that are supported in the actual, compiled implementation. They are obtained from the
- *   intersection of supported and allowed types.
+ * 3. Allowed types are the types for which support is requested (for example, by external configuration).
+ *    Zero or more allowed type sets may be given.
+ *    The supported type set (1) will be limited by set intersection with all allowed type sets.
  *
- * The types are associated with an Op argument. It is also possible to specify a global list of allowed types.
+ * 4. Enabled types are the types that are actually supported.
+ *    These are the required types and the supported, allowed types.
```

Comment:
might be clearer to say 'actually supported in this build'

Reply:
ok. it occurred to me that 'supported' is overloaded in this description - i think i will rename 'supported types' to 'default types' in a future change.